File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/c9fb2d1a92fd15a81ee8f3fdd5b726d0.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef long long int ll;

ll max(ll a, ll b) {
    return (a > b) ? a : b;
}

ll sum(ll *arr, int size) {
    ll s = 0;
    for (int i = 0; i < size; i++) {
        s += arr[i];
    }
    return s;
}

ll rec(int l, int r, int k, ll **dps, ll *sums, int **arrs) {
    ll ans = 0;
    int m = (l + r) / 2;
    if (l + 1 == r) {
        ans = dps[l][k];
        ll sum = 0;
        for (int i = 0; i < k && i < arrs[l][0]; i++) {
            sum += arrs[l][i + 1];
            ans = max(ans, dps[l][k - i - 1] + sum);
        }
    } else {
        for (int i = l; i < m; i++) {
            for (int j = k; j >= arrs[i][0]; j--) {
                dps[i + 1][j] = max(dps[i + 1][j], dps[i][j - arrs[i][0]] + sums[i]);
            }
        }
        ans = rec(m, r, k, dps, sums, arrs);
        for (int i = m; i < r; i++) {
            for (int j = k; j >= arrs[i][0]; j--) {
                dps[i + 1][j] = max(dps[i + 1][j], dps[i][j - arrs[i][0]] + sums[i]);
            }
        }
        ans = max(ans, rec(l, m, k, dps, sums, arrs));
    }
    return ans;
}

void solve() {
    int n, k;
    scanf("%d %d", &n, &k);
    ll *sums = (ll *)malloc(n * sizeof(ll));
    int **arrs = (int **)malloc(n * sizeof(int *));
    for (int i = 0; i < n; i++) {
        int t;
        scanf("%d", &t);
        arrs[i] = (int *)malloc((t + 1) * sizeof(int));
        arrs[i][0] = t;
        for (int j = 1; j <= t; j++) {
            scanf("%d", &arrs[i][j]);
        }
        if (t > k) {
            arrs[i][0] = k;
        }
        sums[i] = sum(arrs[i] + 1, arrs[i][0]);
    }
    ll **dps = (ll **)malloc((n + 1) * sizeof(ll *));
    for (int i = 0; i <= n; i++) {
        dps[i] = (ll *)malloc((k + 1) * sizeof(ll));
        memset(dps[i], 0, (k + 1) * sizeof(ll));
    }
    printf("%lld\n", rec(0, n, k, dps, sums, arrs));
    for (int i = 0; i <= n; i++) {
        free(dps[i]);
    }
    free(dps);
    for (int i = 0; i < n; i++) {
        free(arrs[i]);
    }
    free(arrs);
    free(sums);
}

int main() {
    solve();
    return 0;
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Go/c9fb2d1a92fd15a81ee8f3fdd5b726d0.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
	"strconv"
)

type Solver struct {
	in  *bufio.Reader
	out io.Writer
}

func NewSolver(r io.Reader, w io.Writer) *Solver {
	return &Solver{in: bufio.NewReader(r), out: w}
}

func (s *Solver) readInt() int {
	numStr, _ := s.read()
	num, _ := strconv.Atoi(numStr)
	return num
}

func (s *Solver) read() (string, error) {
	str, err := s.in.ReadString('\n')
	if err != nil {
		return "", err
	}
	return str[:len(str)-1], nil
}

func (s *Solver) write(str string) {
	fmt.Fprint(s.out, str)
}

func (s *Solver) writeLine(str string) {
	fmt.Fprintln(s.out, str)
}

func (s *Solver) solve() {
	n := s.readInt()
	k := s.readInt()

	sums := make([]int, n)
	arrs := make([][]int, n)
	for i := 0; i < n; i++ {
		t := s.readInt()
		arrs[i] = make([]int, t)
		for j := 0; j < t; j++ {
			arrs[i][j] = s.readInt()
		}
		sort.Ints(arrs[i])
		if t > k {
			arrs[i] = arrs[i][:k]
		}
		sums[i] = sum(arrs[i])
	}

	dps := make([][]int, 13)
	for i := 0; i < 13; i++ {
		dps[i] = make([]int, k+1)
	}

	var rec func(int, int, int) int
	rec = func(l, r, dpi int) int {
		ans := 0
		carr := arrs[l]
		dp := dps[dpi]
		if l+1 == r {
			if k < len(carr) {
				ans = max(ans, dp[k])
				sum := 0
				for i := 0; i < min(k, len(carr)); i++ {
					sum += carr[i]
					ans = max(ans, dp[k-i-1]+sum)
				}
			} else {
				ans = max(ans, dp[len(carr)])
			}
		} else {
			m := (l + r) / 2
			copy(dps[dpi+1], dps[dpi])
			for i := l; i < m; i++ {
				v := sums[i]
				k1 := len(arrs[i])
				for j := k; j >= k1; j-- {
					dp[j] = max(dp[j], dp[j-k1]+v)
				}
			}
			ans = max(ans, rec(m, r, dpi+1))
			copy(dps[dpi+1], dps[dpi])
			for i := m; i < r; i++ {
				v := sums[i]
				k1 := len(arrs[i])
				for j := k; j >= k1; j-- {
					dp[j] = max(dp[j], dp[j-k1]+v)
				}
			}
			ans = max(ans, rec(l, m, dpi+1))
		}
		return ans
	}

	s.writeLine(strconv.Itoa(rec(0, n, 0)))
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func sum(arr []int) int {
	s := 0
	for _, num := range arr {
		s += num
	}
	return s
}

func main() {
	solver := NewSolver(os.Stdin, os.Stdout)
	solver.solve()
}
